pkg/controller/cloneset/sync/cloneset_scale.go

```diff
@@ -365,14 +368,68 @@ func getOrGenInstanceID(existingIDs, availableIDs sets.String) string {
 	return id
 }
 
-func choosePodsToDelete(totalDiff int, currentRevDiff int, notUpdatedPods, updatedPods []*v1.Pod) []*v1.Pod {
+// Group the pods according to the number of priority conditions,
+// group the remaining pods into a group, and return after merging
+// This method is a stable sort
+func sortByDeletePriority(priority []appsv1alpha1.CloneSetDeletePriorityWeightTerm, pods []*v1.Pod) []*v1.Pod {
+
+	labelSelectors := make([]labels.Selector, len(priority)+1)
+	podGroups := make([][]*v1.Pod, len(priority)+1)
+	result := make([]*v1.Pod, 0, len(pods))
+
+	sort.SliceStable(priority, func(i, j int) bool {
+		return priority[i].Weight > priority[j].Weight
+	})
+
+	for i, item := range priority {
+		selector, err := metav1.LabelSelectorAsSelector(item.MatchSelector)
+		if err != nil {
+			klog.Errorf("Error DeletePriority.CloneSetDeletePriorityWeightTerm.MatchSelector %s selector: %v", item.MatchSelector, err)
+			return pods
+		}
+		labelSelectors[i] = selector
+	}
+
+	for _, pod := range pods {
+		for i := 0; i <= len(priority); i++ {
+			if labelSelectors[i] != nil && !labelSelectors[i].Matches(labels.Set(pod.Labels)) {
+				continue
+			}
+			podGroups[i] = append(podGroups[i], pod)
+			break
+		}
+	}
+
+	for _, podGroup := range podGroups {
+		result = append(result, podGroup...)
+	}
+	return result
+}
+
+func choosePodsToDelete(cs *appsv1alpha1.CloneSet, totalDiff int, currentRevDiff int, notUpdatedPods, updatedPods []*v1.Pod) []*v1.Pod {
 	choose := func(pods []*v1.Pod, diff int) []*v1.Pod {
 		// No need to sort pods if we are about to delete all of them.
 		if diff < len(pods) {
 			// Sort the pods in the order such that not-ready < ready, unscheduled
 			// < scheduled, and pending < running. This ensures that we delete pods
 			// in the earlier stages whenever possible.
 			sort.Sort(kubecontroller.ActivePods(pods))
+			// Divide the sorted pods into two groups according to whether they accept the request,
```

Comment:
i think it's time to replace  ActivePods  with our own sort function, and insert the delete priority logic inside the less function. sort first with ActivePods, and mess the sorted pods with another sort func is too complex and costly.